pkg/policy/rule.go

```diff
@@ -410,6 +425,29 @@ func (r *rule) canReachIngress(ctx *SearchContext, state *traceState) api.Decisi
 	return api.Undecided
 }
 
+func (r *rule) matches(id uint16, securityIdentity *identity.Identity) bool {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	var ruleMatches bool
+
+	// Rule has already been processed, we can check the caches in the rule.
+	if _, ok := r.processedConsumers[id]; ok {
+		if _, ok := r.localRuleConsumers[id]; ok {
```

Comment:
Maybe check if identity pointer matches that which is in the cache? This would ensure in identity update case that we are always OK. Thoughts?